cmd/cli/uninstall_mesh.go

```diff
@@ -80,70 +81,303 @@ func newUninstallMeshCmd(config *action.Configuration, in io.Reader, out io.Writ
 }
 
 func (d *uninstallMeshCmd) run() error {
-	meshInfoList, _ := getMeshInfoList(d.config, d.clientSet)
-	if len(meshInfoList) == 0 {
-		if d.force {
-			return nil
-		}
-		return errors.Errorf("No osm mesh found")
+	err := d.validateCLIParams()
+	if err != nil {
+		return err
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	ns := settings.Namespace()
+	if !settings.IsManaged() {
+		if !d.force {
+			// print a list of meshes within the cluster for a better user experience
+			fmt.Fprintf(d.out, "\nList of meshes present in the cluster:\n")
+
+			listCmd := &meshListCmd{
+				out:       d.out,
+				config:    d.config,
+				clientSet: d.clientSet,
+				localPort: d.localPort,
+			}
+
+			err := listCmd.run()
+
+			// Unable to list meshes in the cluster
+			if err != nil {
+				return err
+			}
+
+			confirm, err := confirm(d.in, d.out, fmt.Sprintf("\nUninstall OSM [mesh name: %s] in namespace [%s] and/or OSM resources ?", d.meshName, d.meshNamespace), 3)
+			if !confirm || err != nil {
+				return err
+			}
+		}
+
+		_, err = d.client.Run(d.meshName)
+		if err != nil && errors.Cause(err) == helmStorage.ErrReleaseNotFound {
+			fmt.Fprintf(d.out, "No OSM control plane with mesh name [%s] found in namespace [%s]\n", d.meshName, d.meshNamespace)
+
+			if !d.deleteClusterWideResources && !d.deleteNamespace {
+				return err
+			}
+		}
+
+		if err == nil {
+			fmt.Fprintf(d.out, "OSM [mesh name: %s] in namespace [%s] uninstalled\n", d.meshName, d.meshNamespace)
+		}
+	}
+
+	if d.deleteClusterWideResources {
+		var failedDeletions []string
+
+		err = d.uninstallCustomResourceDefinitions()
+		if err != nil {
+			failedDeletions = append(failedDeletions, "CustomResourceDefinitions")
+		}
+
+		err = d.uninstallMutatingWebhookConfigurations()
+		if err != nil {
+			failedDeletions = append(failedDeletions, "MutatingWebhookConfigurations")
+		}
+
+		err = d.uninstallValidatingWebhookConfigurations()
+		if err != nil {
+			failedDeletions = append(failedDeletions, "ValidatingWebhookConfigurations")
+		}
+
+		err = d.uninstallSecrets()
+		if err != nil {
+			failedDeletions = append(failedDeletions, "Secrets")
+		}
+
+		if len(failedDeletions) != 0 {
+			return errors.Errorf("Failed to completely delete the following OSM resource types: %+v", failedDeletions)
+		}
+	}
+
+	if !settings.IsManaged() {
+		if d.deleteNamespace {
+			if err = d.clientSet.CoreV1().Namespaces().Delete(ctx, d.meshNamespace, v1.DeleteOptions{}); err != nil {
+				if k8sApiErrors.IsNotFound(err) {
+					fmt.Fprintf(d.out, "OSM namespace [%s] not found\n", d.meshNamespace)
+					return nil
+				}
+				return errors.Errorf("Error occurred while deleting OSM namespace [%s] - %v", d.meshNamespace, err)
+			}
+			fmt.Fprintf(d.out, "OSM namespace [%s] deleted successfully\n", d.meshNamespace)
+		}
+	} else {
+		if d.deleteNamespace {
+			fmt.Fprintf(d.out, "OSM namespace [%s] CANNOT be deleted in a managed environment\n", d.meshNamespace)
+		}
+	}
```

Comment:
Just a suggestion :)

```suggestion
	if d.deleteNamespace {
		 if !settings.IsManaged() {
			if err = d.clientSet.CoreV1().Namespaces().Delete(ctx, d.meshNamespace, v1.DeleteOptions{}); err != nil {
				if k8sApiErrors.IsNotFound(err) {
					fmt.Fprintf(d.out, "OSM namespace [%s] not found\n", d.meshNamespace)
					return nil
				}
				return errors.Errorf("Error occurred while deleting OSM namespace [%s] - %v", d.meshNamespace, err)
			}
			fmt.Fprintf(d.out, "OSM namespace [%s] deleted successfully\n", d.meshNamespace)
		} else {
			fmt.Fprintf(d.out, "OSM namespace [%s] CANNOT be deleted in a managed environment\n", d.meshNamespace)
		}
	}
```